File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/sixplay/SixPlayExtractor.java
package lan.dk.podcastserver.manager.worker.sixplay;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.jayway.jsonpath.DocumentContext;
import com.jayway.jsonpath.TypeRef;
import com.mashape.unirest.http.HttpResponse;
import io.vavr.collection.List;
import io.vavr.control.Option;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.manager.downloader.DownloadingItem;
import lan.dk.podcastserver.manager.worker.Extractor;
import lan.dk.podcastserver.service.HtmlService;
import lan.dk.podcastserver.service.JsonService;
import lan.dk.podcastserver.service.M3U8Service;
import lan.dk.podcastserver.service.UrlService;
import lombok.*;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.select.Elements;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import static io.vavr.API.Try;
import static org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE;

/**
 * Created by kevin on 26/12/2017
 */
@Slf4j
@Component
@Scope(SCOPE_PROTOTYPE)
@RequiredArgsConstructor
public class SixPlayExtractor implements Extractor {

    private static final TypeRef<List<M6PlayItem>> TYPE_ITEMS = new TypeRef<List<M6PlayItem>>(){};
    private static final String ITEMS_EXTRACTOR = "video.currentVideo.clips[*]";

    private final HtmlService htmlService;
    private final JsonService jsonService;
    private final M3U8Service m3U8Service;
    private final UrlService urlService;

    @Override
    public DownloadingItem extract(Item item) {
        List<String> videos = htmlService.get(item.getUrl())
                .map(d -> d.select("script"))
                .map(this::extractUrl)
                .getOrElse(List::empty);

        return DownloadingItem.builder()
                .item(item)
                .urls(videos)
                .filename(getFileName(item))
                .build();
    }

    private List<String> extractUrl(Elements script) {
        return extractJson(script)
                .map(JsonService.to(ITEMS_EXTRACTOR, TYPE_ITEMS))
                .getOrElse(List::empty)
                .flatMap(this::keepBestQuality)
                .map(M6PlayAssets::getFull_physical_path);
    }

    private Option<DocumentContext> extractJson(Elements elements) {
        return SixPlayUpdater.getRoot6Play(elements).map(jsonService::parse);
    }

    private Option<M6PlayAssets> keepBestQuality(M6PlayItem item) {
        List<M6PlayAssets> assets = item.getAssets();

        return assets
                .filter(i -> !i.getProtocol().equalsIgnoreCase("primetime"))
                .find(i -> StringUtils.contains(i.getVideo_quality(), "sd3"))
                .flatMap(this::transformSd3Url)
                .orElse(() -> assets.find(i -> "usp_hls_h264".equalsIgnoreCase(i.getType())))
                .orElse(() -> assets.find(i -> "hq".equalsIgnoreCase(i.getVideo_quality())))
                .orElse(() -> assets.find(i -> "hd".equalsIgnoreCase(i.getVideo_quality())))
                .orElse(() -> assets.find(i -> "sd".equalsIgnoreCase(i.getVideo_quality())));
    }

    private Option<M6PlayAssets> transformSd3Url(M6PlayAssets asset) {
        String realURL = urlService.getRealURL(asset.getFull_physical_path());

        return Try(() -> urlService.get(realURL)
                .header(UrlService.USER_AGENT_KEY, UrlService.USER_AGENT_MOBILE)
                .asString())
                .map(HttpResponse::getRawBody)
                .flatMap(is -> m3U8Service.findBestQuality(is).toTry())
                .map(url -> urlService.addDomainIfRelative(realURL, url))
                .map(url -> new M6PlayAssets(asset.getVideo_quality(), url, asset.getType(), asset.getProtocol()))
                .toOption();
    }

    @Override
    public String getFileName(Item item) {
        return Extractor.super.getFileName(item) + ".mp4";
    }

    @Override
    public Integer compatibility(String url) {
        return SixPlayUpdater.isFrom6Play(url);
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class M6PlayItem {
        @Getter @Setter private List<M6PlayAssets> assets;
    }

    @AllArgsConstructor @NoArgsConstructor
    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class M6PlayAssets {
        @Setter @Getter private String video_quality;
        @Setter @Getter private String full_physical_path;
        @Setter @Getter private String type;
        @Setter @Getter private String protocol;
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/sixplay/SixPlayExtractor.kt
package com.github.davinkevin.podcastserver.manager.worker.sixplay

import arrow.core.Option
import arrow.core.getOrElse
import arrow.core.orElse
import arrow.core.toOption
import arrow.syntax.collections.firstOption
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.github.davinkevin.podcastserver.manager.downloader.DownloadingItem
import com.github.davinkevin.podcastserver.manager.worker.Extractor
import com.github.davinkevin.podcastserver.service.M3U8Service
import com.github.davinkevin.podcastserver.service.UrlService
import com.github.davinkevin.podcastserver.utils.MatcherExtractor.Companion.from
import com.github.davinkevin.podcastserver.utils.k
import com.jayway.jsonpath.TypeRef
import lan.dk.podcastserver.entity.Item
import lan.dk.podcastserver.service.JsonService
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

/**
 * Created by kevin on 26/12/2017
 */
@Component
@Scope(SCOPE_PROTOTYPE)
class SixPlayExtractor(val jsonService: JsonService, val m3U8Service: M3U8Service, val urlService: UrlService) : Extractor {

    override fun extract(item: Item): DownloadingItem {

        val m = URL_EXTRACTOR.on(item.url)

        val list = (m.group(1).k() to m.group(2).k())
                .map { toJsonUrl(it) }
                .flatMap { jsonService.parseUrl(it).k() }
                .map { it.read(ITEMS_EXTRACTOR, TYPE_ITEMS) }
                .getOrElse { throw RuntimeException("No element founds for ${item.id} at url ${item.url}") }

        val urls: List<String> = list
                .flatMap { keepBestQuality(it).toList() }
                .map { it.full_physical_path }
                .filterNotNull()

        return DownloadingItem(
                item,
                urls,
                getFileName(item),
                null
        )
    }

    private fun keepBestQuality(item: M6PlayItem): Option<M6PlayAssets> {
        val a = item.assets
                .filterNot { it.protocol == "primetime" }

        return a
                .firstOption{ (it.video_quality ?: "").contains("sd1") }.flatMap { transformSD1Ism(it) }
                .orElse { a.firstOption{ (it.video_quality ?: "").contains("sd3") }.flatMap { transformSd3Url(it) } }
                .orElse { a.firstOption { i -> "usp_hls_h264" == i.type } }
                .orElse { a.firstOption { i -> "hq" == i.video_quality } }
                .orElse { a.firstOption { i -> "hd" == i.video_quality } }
                .orElse { a.firstOption { i -> "sd" == i.video_quality } }
    }

    private fun transformSd3Url(asset: M6PlayAssets): Option<M6PlayAssets> = transformUrl(asset)
    private fun transformSD1Ism(asset: M6PlayAssets): Option<M6PlayAssets> = transformUrl(asset) { toFullQualityPlaylistUrl(it) }

    private fun transformUrl(asset: M6PlayAssets, urlMapper: (String) -> String = { it }): Option<M6PlayAssets> =
            urlService.getRealURL(asset.full_physical_path!!).toOption()
                    .map { urlMapper(it) }
                    .map { urlService.get(it)
                            .header(UrlService.USER_AGENT_KEY, UrlService.USER_AGENT_MOBILE)
                            .asString() }
                    .map { it.rawBody }
                    .flatMap { m3U8Service.findBestQuality(it).k() }
                    .map { urlService.addDomainIfRelative(urlService.getRealURL(asset.full_physical_path!!), it) }
                    .map { M6PlayAssets(asset.video_quality, it, asset.type, asset.protocol) }

    override fun getFileName(item: Item) = super.getFileName(item) + ".mp4"
    override fun compatibility(url: String?) = SixPlayUpdater.isFrom6Play(url)

    private fun toFullQualityPlaylistUrl(t: String): String {
        val base = t.substringBeforeLast("/")
        val name = base.substringAfterLast("/").replace(".ism", ".m3u8")
        return "$base/$name"
    }

    companion object {

        private const val ITEMS_EXTRACTOR = "clips[*]"
        private const val INFO_URL = "https://pc.middleware.6play.fr/6play/v2/platforms/m6group_web/services/6play/videos/%s_%s?with=clips&csa=5"
        private val TYPE_ITEMS = object : TypeRef<List<M6PlayItem>>(){}
        private val URL_EXTRACTOR = from("^.+6play\\.fr/.+-([a-z])_([0-9]*)")

        private fun expandType(type: String): String = when(type) {
            "p" -> "playlist"
            "c" -> "clip"
            else -> throw RuntimeException("Invalid type \"$type\" for 6play item")
        }

        private fun toJsonUrl(p: Pair<String, String>): String {
            return String.format(INFO_URL, expandType(p.first), p.second)
        }
    }
}

@JsonIgnoreProperties(ignoreUnknown = true)
private class M6PlayItem {
    var assets: List<M6PlayAssets> = listOf()
}

@JsonIgnoreProperties(ignoreUnknown = true)
private class M6PlayAssets(
        var video_quality: String? = null,
        var full_physical_path: String? = null,
        var type: String? = null,
        var protocol: String? = null
)

private infix fun <T, U> Option<T>.to(other: Option<U>): Option<Pair<T, U>> = this.flatMap { first -> other.map { first to it } }
